core/syncstatus/objectsyncstatus/syncstatus.go

```diff
@@ -0,0 +1,237 @@
+	watchers  map[string]struct{}
+
+	treeStatusBuf []treeStatus
+
+	updateIntervalSecs int
+	updateTimeout      time.Duration
+}
+
+func NewSyncStatusService() StatusService {
+	return &syncStatusService{
+		treeHeads: map[string]treeHeadsEntry{},
+		watchers:  map[string]struct{}{},
+	}
+}
+
+func (s *syncStatusService) Init(a *app.App) (err error) {
+	sharedState := a.MustComponent(spacestate.CName).(*spacestate.SpaceState)
+	s.updateIntervalSecs = syncUpdateInterval
+	s.updateTimeout = syncTimeout
+	s.spaceId = sharedState.SpaceId
+	s.configuration = app.MustComponent[nodeconf.NodeConf](a)
+	s.storage = app.MustComponent[spacestorage.SpaceStorage](a)
+	s.periodicSync = periodicsync.NewPeriodicSync(
+		s.updateIntervalSecs,
+		s.updateTimeout,
+		s.update,
+		log)
+	return
+}
+
+func (s *syncStatusService) Name() (name string) {
+	return syncstatus.CName
+}
+
+func (s *syncStatusService) SetUpdateReceiver(updater UpdateReceiver) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.updateReceiver = updater
+}
+
+func (s *syncStatusService) Run(ctx context.Context) error {
+	s.periodicSync.Run()
+	return nil
+}
+
+func (s *syncStatusService) HeadsChange(treeId string, heads []string) {
+	s.Lock()
+	defer s.Unlock()
+
+	var headsCopy []string
+	headsCopy = append(headsCopy, heads...)
+
+	s.treeHeads[treeId] = treeHeadsEntry{
+		heads:      headsCopy,
+		syncStatus: StatusNotSynced,
+	}
+}
+
+func (s *syncStatusService) update(ctx context.Context) (err error) {
+	s.treeStatusBuf = s.treeStatusBuf[:0]
+
+	s.Lock()
+	if s.updateReceiver == nil {
+		s.Unlock()
+		return
+	}
+	for treeId := range s.watchers {
+		// that means that we haven't yet got the status update
+		treeHeads, exists := s.treeHeads[treeId]
+		if !exists {
+			err = fmt.Errorf("treeHeads should always exist for watchers")
+			s.Unlock()
+			return
+		}
+		s.treeStatusBuf = append(s.treeStatusBuf, treeStatus{treeId, treeHeads.syncStatus})
+	}
+	s.Unlock()
+	s.updateReceiver.UpdateNodeStatus()
+	for _, entry := range s.treeStatusBuf {
+		err = s.updateReceiver.UpdateTree(ctx, entry.treeId, entry.status)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+func (s *syncStatusService) HeadsReceive(senderId, treeId string, heads []string) {
+	s.Lock()
+	defer s.Unlock()
+
+	curTreeHeads, ok := s.treeHeads[treeId]
+	if !ok || curTreeHeads.syncStatus == StatusSynced {
+		return
+	}
+
+	// checking if other node is responsible
+	if len(heads) == 0 || !s.isSenderResponsible(senderId) {
+		return
+	}
+
+	// checking if we received the head that we are interested in
+	for _, head := range heads {
+		if idx, found := slices.BinarySearch(curTreeHeads.heads, head); found {
+			curTreeHeads.heads[idx] = ""
+		}
+	}
+	curTreeHeads.heads = slice.DiscardFromSlice(curTreeHeads.heads, func(h string) bool {
+		return h == ""
+	})
+	if len(curTreeHeads.heads) == 0 {
+		curTreeHeads.syncStatus = StatusSynced
+	}
+	s.treeHeads[treeId] = curTreeHeads
+}
+
+func (s *syncStatusService) Watch(treeId string) (err error) {
```

Comment:
It is important to clear statuses of objects when we call SyncAll with responsible peers (before the method was called something like `RemoveAllExcept...`)